### PR TITLE
#695 Move Code List SPARQL Queries to CodeListInspector

### PR DESCRIPTION
**Pull request by nimshi89**
#695

transferred the following queries to cached properties:
- select_codelist_csv_url.sparql (code_list_inspector.py) - reworked to link between the csv_url and conceptSchemeUrl
- select_codelist_cols_by_csv_url.sparql (csvw_state.py) - replaced with a generic columns query in csvw_state

Adds a get_catalog_metadata API function to code_list_inspector.py which uses the results from the select_codelist_csv_url.sparql cached property.

Adds unit test coverage for new functions/properties

Class has been renamed to  CodeListInspector

### CODE REVIEW

File: src/csvcubed/utils/sparql_handler/code_list_inspector.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    CodeListTableIdentifers,
+    ColumnDefinition,
+)
+from csvcubed.utils.iterables import first
+from csvcubed.utils.sparql_handler.csvw_state import CsvWState
+
+
+@dataclass
+class CodeListInspector:
+    """Fills this in"""
```
Comment (CharlesRendle) [inline src/csvcubed/utils/sparql_handler/code_list_inspector.py:19]: Needs a comment

File: src/csvcubed/utils/sparql_handler/code_list_inspector.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+        return identifiers
+
+    def get_csvw_catalog_metadata(self) -> CatalogMetadataResult:
+        """This function will access the catalogmetadataResult and compares the concept_sceme_url and returns the relevant value"""
+        the_list = self.csvw_state.catalog_metadata
+
+        concept_scheme_url = self._code_list_table_identifiers[0].concept_scheme_url
+
+        for x in the_list:
```
Comment (CharlesRendle) [inline src/csvcubed/utils/sparql_handler/code_list_inspector.py:73]: We probably want to rename `x` and `the_list` to something else

File: src/csvcubed/utils/sparql_handler/data_cube_state.py [hunk 3]
```diff
@@ -1,57 +1,50 @@
...
-    """
-    Private utility functions.
-    """
-
-    def _get_value_for_key(self, key: str, dict: Dict[str, T]) -> T:
-        maybe_value = dict.get(key)
-        if maybe_value is None:
-            raise KeyError(f"Could not find the definition for key '{key}'")
-        return maybe_value
+    """todo: Describe the class"""
```
Comment (CharlesRendle) [inline src/csvcubed/utils/sparql_handler/data_cube_state.py:39]: Needs a comment

File: src/csvcubed/utils/sparql_handler/data_cube_state.py [hunk 4]
```diff
@@ -1,57 +1,50 @@
...
-        Map of csv_url to the list of column definitions for the given CSV file.
-        """
-        results = select_column_definitions(self.csvw_state.rdf_graph)
-        return group_by(results, lambda r: r.csv_url)
+    # Private cached properties.
 
     @cached_property
     def _units(self) -> Dict[str, UnitResult]:
-        """ """
+        """todo: Fill me in"""
```
Comment (CharlesRendle) [inline src/csvcubed/utils/sparql_handler/data_cube_state.py:47]: Needs a comment

File: src/csvcubed/utils/sparql_handler/data_cube_state.py [hunk 5]
```diff
@@ -125,33 +119,22 @@ def _detect_shape_for_cube(
...
-        """
-        result: List[ColumnDefinition] = self._get_value_for_key(
-            csv_url, self._column_definitions
-        )
-        return result
+    # Public getters for the cached properties.
 
     def get_unit_for_uri(self, uri: str) -> Optional[UnitResult]:
-        """ """
+        """todo: Need to fill this in."""
```
Comment (CharlesRendle) [inline src/csvcubed/utils/sparql_handler/data_cube_state.py:125]: Needs a comment

File: src/csvcubed/utils/sparql_handler/data_cube_state.py [hunk 6]
```diff
@@ -125,33 +119,22 @@ def _detect_shape_for_cube(
...
         return self._units.get(uri)
 
     def get_units(self) -> List[UnitResult]:
-        """ """
+        """todo: Need to fill this in."""
```
Comment (CharlesRendle) [inline src/csvcubed/utils/sparql_handler/data_cube_state.py:129]: Needs a comment

File: src/csvcubed/utils/sparql_handler/sparql_queries/select_codelist_csv_url.sparql [hunk 7]
```diff
@@ -3,8 +3,8 @@ PREFIX csvw: <http://www.w3.org/ns/csvw#>
 PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
 
 
-SELECT DISTINCT ?conceptScheme ?tableUrl
-WHERE {   
+SELECT DISTINCT ?conceptScheme ?tableUrl ?conceptSchemeUrlStr
```
Comment (CharlesRendle) [inline src/csvcubed/utils/sparql_handler/sparql_queries/select_codelist_csv_url.sparql:6]: Pretty sure I added this `?conceptSchemeUrlStr` last week while playing around with the query. We should make sure we definitely dont need it

File: tests/test-cases/cli/inspect/itis-industry-multiple-skos-inscheme.csv-metadata2.json [hunk 8]
```diff
@@ -0,0 +1,124 @@
+{
```
Comment (CharlesRendle) [inline tests/test-cases/cli/inspect/itis-industry-multiple-skos-inscheme.csv-metadata2.json:1]: Can we give this a better title which indicates what is different/why we need it

File: tests/test-cases/cli/inspect/itis-industry.csv-metadata.json [hunk 9]
```diff
@@ -71,7 +71,7 @@
     },
     "rdfs:seeAlso": [
         {
-            "@id": "http://gss-data.org.uk/data/gss_data/trade/ons-international-trade-in-services#scheme/itis-industry/dataset",
+            "@id": "http://gss-data.org.uk/data/gss_data/trade/ons-international-trade-in-services#scheme/itis-industry",
```
Comment (CharlesRendle) [inline tests/test-cases/cli/inspect/itis-industry.csv-metadata.json:74]: Why was this change needed?

File: tests/unit/cli/inspect/test_code_list_inspector.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+from csvcubed.utils.tableschema import CsvwRdfManager
+from tests.unit.test_baseunit import get_test_cases_dir
+
+_test_case_base_dir = (
+    get_test_cases_dir() / "cli" / "inspect" / "pivoted-multi-measure-dataset"
+)
+
+
+def test_code_list_table_identifiers():
+    """the test checks for the codelist idetifiers are available but
```
Comment (CharlesRendle) [inline tests/unit/cli/inspect/test_code_list_inspector.py:14]: I dont think this is correct

File: tests/unit/cli/inspect/test_code_list_inspector.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    assert len(code_list_identifiers) == 1
+    assert code_list_identifiers[0] == CodeListTableIdentifers(
+        csv_url="some-dimension.csv", concept_scheme_url="some-dimension.csv#code-list"
+    )
+    # it should have one value returned
+
+
+# make a fail test for the one above
+def test_code_list_table_identifiers_error():
+    """the test checks for the codelist idetifiers are available but
```
Comment (CharlesRendle) [inline tests/unit/cli/inspect/test_code_list_inspector.py:33]: Or this one

File: tests/unit/cli/inspect/test_code_list_inspector.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+    with pytest.raises(KeyError) as exception:
+        _ = code_list_inspector._code_list_table_identifiers
+
+    assert (
+        "Found multiple skos:inScheme columns in 'itis-industry-multiple-skos-inscheme.csv'."
+    ) in str(exception.value)
+
+
+def test_get_csvw_catalog_metadata():
```
Comment (CharlesRendle) [inline tests/unit/cli/inspect/test_code_list_inspector.py:55]: Needs a comment

File: tests/unit/cli/inspect/test_code_list_inspector.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+
+    code_list_inspector = CodeListInspector(rdf_manager.csvw_state)
+
+    the_result = code_list_inspector.get_csvw_catalog_metadata()
+
+    assert isinstance(the_result, CatalogMetadataResult)
+    assert the_result.dataset_uri == "some-dimension.csv#code-list"
+
+
+def test_get_csvw_catalog_metadata_error():
```
Comment (CharlesRendle) [inline tests/unit/cli/inspect/test_code_list_inspector.py:69]: Needs a comment

File: tests/unit/cli/inspect/test_code_list_inspector.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
+    with pytest.raises(ValueError) as exception:
+        _ = code_list_inspector.get_csvw_catalog_metadata()
+
+    assert (
+        "None of the results can be associated with the http://gss-data.org.uk/data/gss_data/trade/ons-international-trade-in-services#scheme/it-industry"
+    ) in str(exception.value)
+
+
+def test_get_table_identifiers_for_concept_scheme_error():
```
Comment (CharlesRendle) [inline tests/unit/cli/inspect/test_code_list_inspector.py:90]: Needs a comment

File: tests/unit/utils/sparqlhandler/test_code_list_state.py [hunk 15]
```diff
@@ -0,0 +1,31 @@
+from csvcubed.utils.sparql_handler.code_list_inspector import CodeListInspector
```
Comment (CharlesRendle) [inline tests/unit/utils/sparqlhandler/test_code_list_state.py:1]: Im pretty sure we dont need this test file anymore as we populated `test_code_list_inspector.py` instead

File: src/csvcubed/utils/sparql_handler/code_list_inspector.py (hunk 2 above)
Comment (nimshi89) [reply src/csvcubed/utils/sparql_handler/code_list_inspector.py:73]: done

File: tests/test-cases/cli/inspect/itis-industry-multiple-skos-inscheme.csv-metadata2.json (hunk 8 above)
Comment (nimshi89) [reply tests/test-cases/cli/inspect/itis-industry-multiple-skos-inscheme.csv-metadata2.json:1]: done

**Review by NickPapONS — COMMENTED** [review]
Looks good, makes sense to me! There are a few differences in the way cached properties are implemented compared to how we did it in the other tickets with myself and Sarah (data_cube_state or csvw_state), but I think these cached properties mostly work independently between code lists and data cubes. There will be quite a bit of conflict resolution needed when we merge all these 3 cached properties tickets together, but that's a natural part of splitting up the task the way we did. I don't see anything here that I think we wouldn't want to keep anyway.

File: src/csvcubed/utils/sparql_handler/csvw_state.py [hunk 16]
```diff
@@ -25,6 +30,18 @@ class CsvWState:
     def __post_init__(self):
         self.primary_graph_uri = path_to_file_uri_for_rdflib(self.csvw_json_path)
 
+    """Private Functions"""
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/csvw_state.py:33]: There are public things below this comment.

File: src/csvcubed/utils/sparql_handler/csvw_state.py [hunk 17]
```diff
@@ -25,6 +30,18 @@ class CsvWState:
...
+
+    @cached_property
+    def column_definitions(self) -> Dict[str, List[ColumnDefinition]]:
```
[the remark this answers is not in the document]
Comment (robons) [reply src/csvcubed/utils/sparql_handler/csvw_state.py:36]: I'm not aware of any other queries which could make use of this. But it may well be worth considering if any changes you've been making could make use of this information.

File: src/csvcubed/utils/sparql_handler/data_cube_state.py [hunk 18]
```diff
@@ -125,33 +119,22 @@ def _detect_shape_for_cube(
...
+    # Public getters for the cached properties.
 
     def get_unit_for_uri(self, uri: str) -> Optional[UnitResult]:
-        """ """
+        """The function returns a single unit from the uri if there is any."""
         return self._units.get(uri)
 
     def get_units(self) -> List[UnitResult]:
-        """ """
+        """This function gets the value of the unit and returns it as a list."""
```
[the remark this answers is not in the document]
Comment (robons) [reply src/csvcubed/utils/sparql_handler/data_cube_state.py:129]: Yeah, it's just been moved into the `CsvWInspector` since it's useful in both `CodeListInspector` and `DataCubeInspector` and describes things at the CSV-W level.

File: tests/behaviour/steps/inspectcli.py [hunk 19]
```diff
@@ -129,7 +129,7 @@ def step_impl(context):
         context.csvw_metadata_rdf_graph,
         context.csvw_metadata_json_path,
     )
-    code_list_state = CodeListState(csvw_state)
+    code_list_state = CodeListInspector(csvw_state)
```
Comment (robons) [inline tests/behaviour/steps/inspectcli.py:132]: Could you rename `code_list_state` to `code_list_inspector` as well?

File: tests/test-cases/cli/inspect/itis-industry-no-skos-inscheme.csv-metadata.json [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+                "propertyUrl": "rdfs:comment"
+            },
+            {
+                "virtual": true,
+                "propertyUrl": "rdf:type",
+                "valueUrl": "skos:Concept"
+            },
+            {
+                "virtual": true,
+                "propertyUrl": "skos:inScheme",
```
Comment (robons) [inline tests/test-cases/cli/inspect/itis-industry-no-skos-inscheme.csv-metadata.json:58]: It looks like there's a `skos:inScheme` column in here which disagrees with the name of the file.

File: tests/test-cases/cli/inspect/itis-industry.csv-metadata.json (hunk 9 above)
Comment (robons) [reply tests/test-cases/cli/inspect/itis-industry.csv-metadata.json:74]: IIRC, this was because this code list is a very old one where the `dcat:Dataset` and `skos:ConceptScheme` were seperate resources. This is not how csvcubed generates outputs at the moment, so we've altered the input to the style we're expecting.

File: tests/unit/cli/inspect/test_code_list_inspector.py [hunk 21]
```diff
@@ -0,0 +1,115 @@
+import pytest
+
+from csvcubed.models.sparqlresults import CatalogMetadataResult, CodeListTableIdentifers
+from csvcubed.utils.sparql_handler.code_list_inspector import CodeListInspector
+from csvcubed.utils.tableschema import CsvwRdfManager
+from tests.unit.test_baseunit import get_test_cases_dir
+
+_test_case_base_dir = (
```
Comment (robons) [inline tests/unit/cli/inspect/test_code_list_inspector.py:8]: I think you should alter the base dir to be up to the `inspect` dir so it can be referenced in all of the tests.

File: tests/unit/cli/inspect/test_code_list_inspector.py [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+    code_list_inspector = CodeListInspector(rdf_manager.csvw_state)
+
+    concept_scheme_url = "http://gss-data.org.uk/data/gss_data/trade/ons-international-trade-in-services#scheme/itis-industry"
+
+    with pytest.raises(KeyError) as exception:
+        _ = code_list_inspector.get_table_identifiers_for_concept_scheme(
+            concept_scheme_url
+        )
+    assert (
+        "Could not find code list table identifiers for ConceptSchem URL: 'http://gss-data.org.uk/data/gss_data/trade/ons-international-trade-in-services#scheme/itis-industry'"
```
Comment (robons) [inline tests/unit/cli/inspect/test_code_list_inspector.py:114]: s/ConceptSchem/ConceptScheme/

**Review by robons — CHANGES_REQUESTED** [review]

File: src/csvcubed/utils/sparql_handler/code_list_inspector.py [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+            for (csv_url, columns) in self.csvw_state.column_definitions.items()
+        ]
+
+        return [i for i in table_identifiers if i is not None]
+
+    def get_table_identifiers_for_concept_scheme(
+        self, concept_scheme_url: str
+    ) -> CodeListTableIdentifers:
+        """This functions check the CodeListTableIdentifers and returns the value thats concept_scheme_url mathces with the given argument."""
+        identifiers = first(
```
[the remark this answers is not in the document]
Comment (nimshi89) [reply src/csvcubed/utils/sparql_handler/code_list_inspector.py:66]: well Rob, I don't know how to break this to you. But that is good as it is ;)

File: src/csvcubed/utils/sparql_handler/code_list_inspector.py [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+            for (csv_url, columns) in self.csvw_state.column_definitions.items()
+        ]
+
+        # If the csv file doesn't have a code list in it, don't include it in the code list identifiers returned.
+        return [i for i in table_identifiers if i is not None]
+
+    def get_table_identifiers_for_concept_scheme(
+        self, concept_scheme_url: str
+    ) -> CodeListTableIdentifers:
+        """This function check the CodeListTableIdentifers and returns the value thats concept_scheme_url mathces with the given argument."""
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/code_list_inspector.py:73]: I'd be tempted to describe this as:
```
Returns the table identifiers (csv URL & concept scheme URL) for a given concept scheme URL. 

Raises a KeyError if it cannot be found.
```

File: src/csvcubed/utils/sparql_handler/code_list_inspector.py [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+        """This function will access the catalogmetadataResult and compares the concept_sceme_url and returns the relevant value"""
+        catalog_mdata_results = self.csvw_state.catalog_metadata
+
+        result = first(
+            catalog_mdata_results, lambda i: i.dataset_uri == concept_scheme_url
+        )
+
+        if result is None:
+            raise ValueError(
+                f"None of the results can be associated with the {concept_scheme_url}"
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/code_list_inspector.py:98]: I'd put `concept scheme URL '{concept_scheme_url}'` here to make it clear what it is.

tbh I'd really be tempted to say `Can not find Catalogue Meatadata associated with the concept scheme URL '{concept_scheme_url}'.`.

I think this should be a `KeyError`, not a `ValueError`.

File: src/csvcubed/utils/sparql_handler/code_list_inspector.py [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+            raise KeyError(
+                f"Could not find code list table identifiers for ConceptScheme URL: '{concept_scheme_url}'"
+            )
+
+        return identifiers
+
+    def get_catalog_metadata_for_concept_scheme(
+        self, concept_scheme_url: str
+    ) -> CatalogMetadataResult:
+        """This function will access the catalogmetadataResult and compares the concept_sceme_url and returns the relevant value"""
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/code_list_inspector.py:89]: I'd describe this as:
```
Returns the Catalogue Metadata for a given ConceptScheme. 

Raises a KeyError if it cannot be found.
```

File: src/csvcubed/utils/sparql_handler/csvw_state.py [hunk 27]
```diff
@@ -1,19 +1,24 @@
...
+from csvcubed.utils.iterables import group_by
 from csvcubed.utils.sparql_handler.sparql import path_to_file_uri_for_rdflib
 from csvcubed.utils.sparql_handler.sparqlquerymanager import (
     ask_is_csvw_code_list,
     ask_is_csvw_qb_dataset,
+    select_column_definitions,
     select_csvw_catalog_metadata,
 )
 
+T = TypeVar("T")
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/csvw_state.py:20]: I don't think we need this type variable anymore.

File: src/csvcubed/utils/sparql_handler/csvw_state.py [hunk 28]
```diff
@@ -57,3 +70,12 @@ def get_primary_catalog_metadata(self) -> CatalogMetadataResult:
         raise KeyError(
             f"Could not find catalog metadata in primary graph '{self.primary_graph_uri}'."
         )
+
+    def get_column_definitions_for_csv(self, csv_url: str) -> List[ColumnDefinition]:
+        """
+        Returns the `ColumnDefinition`s for a given csv file.
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/csvw_state.py:76]: Can you mention that it raises a key error if it can't be found (inside the docstring comment)?

File: src/csvcubed/utils/sparql_handler/data_cube_state.py [hunk 29]
```diff
@@ -1,57 +1,50 @@
+"""
+Data Cube State
+---------------
+
+Provides access to inspect the contents of an rdflib.ConjunctiveGraph containing
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/data_cube_state.py:5]: Can you not mention conjunctive graphs here?

File: src/csvcubed/utils/sparql_handler/data_cube_state.py [hunk 30]
```diff
@@ -1,57 +1,50 @@
...
 from csvcubed.utils.sparql_handler.sparqlquerymanager import (
-    select_column_definitions,
     select_csvw_dsd_qube_components,
     select_data_set_dsd_and_csv_url,
     select_is_pivoted_shape_for_measures_in_data_set,
     select_units,
 )
+from csvcubed.utils.uri import looks_like_uri
 
 T = TypeVar("T")
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/data_cube_state.py:34]: Do we need this type var anymore?

File: src/csvcubed/utils/sparql_handler/data_cube_state.py (hunk 4 above)
Comment (robons) [reply src/csvcubed/utils/sparql_handler/data_cube_state.py:47]: Doesn't look like this has been completed to me.

File: src/csvcubed/utils/sparql_handler/data_cube_state.py [hunk 31]
```diff
@@ -125,33 +119,22 @@ def _detect_shape_for_cube(
...
         return self._units.get(uri)
 
     def get_units(self) -> List[UnitResult]:
-        """ """
+        """This function gets the value of the unit and returns it as a list."""
         return list(self._units.values())
 
     def get_cube_identifiers_for_csv(self, csv_url: str) -> CubeTableIdentifiers:
         """
         Getter for data_set_dsd_and_csv_url_for_csv_url cached property.
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/data_cube_state.py:134]: Can you mention it throws a KeyError if it can't find it? (and do the same in the other places in the file which use `get_from_dict_ensure_exists`?

File: src/csvcubed/utils/sparql_handler/data_cube_state.py [hunk 32]
```diff
@@ -174,7 +158,10 @@ def get_dsd_qube_components_for_csv(self, csv_url: str) -> QubeComponentsResult:
         """
         Getter for DSD Qube Components cached property.
         """
-        return self._get_value_for_key(csv_url, self._dsd_qube_components)
+        return get_from_dict_ensure_exists(self._dsd_qube_components, csv_url)
 
     def get_shape_for_csv(self, csv_url: str) -> CubeShape:
-        return self._get_value_for_key(csv_url, self._cube_shapes)
+        """
+        Getter for CubeShape.
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/data_cube_state.py:165]: Can you say something like:

```
Returns the shape of the cube stored in a given CSV file. 

Raises a KeyError if the csv_url is not associated with a cube.
```

File: tests/helpers/inspectors_cache.py [hunk 33]
```diff
@@ -29,3 +31,12 @@ def get_data_cube_inspector(csvw_json_path: Path) -> DataCubeState:
         _data_cube_state_cache[path_str] = DataCubeState(rdf_manager.csvw_state)
 
     return _data_cube_state_cache[path_str]
+
+
+def get_code_list_inspector(csvw_json_path: Path) -> CodeListInspector:
```
Comment (robons) [inline tests/helpers/inspectors_cache.py:36]: Nice work

**Review by robons — COMMENTED** [review]
Almost there, just tidy up those comments and I think we're pretty much there.

File: src/csvcubed/utils/sparql_handler/csvw_state.py [hunk 34]
```diff
@@ -57,3 +68,13 @@ def get_primary_catalog_metadata(self) -> CatalogMetadataResult:
         raise KeyError(
             f"Could not find catalog metadata in primary graph '{self.primary_graph_uri}'."
         )
+
+    def get_column_definitions_for_csv(self, csv_url: str) -> List[ColumnDefinition]:
+        """
+        Returns the `ColumnDefinition`s for a given csv file, raises a KeyError if the csv_url
+        is not associated with a ColumnDefinition.
```
Comment (robons) [inline src/csvcubed/utils/sparql_handler/csvw_state.py:75]: A *list* of column definitions.

**Review by robons — APPROVED** [review]
